xla/service/gpu/fusions/transpose_mlir.cc

```diff
@@ -75,7 +75,52 @@ using mlir_converter::ApplyAffineMap;
 using mlir_converter::CallTargetProvider;
 using mlir_converter::PartitionedComputation;
 
-Tiling ComputeTransposeTiling(const TransposeDescription& tiled_transpose) {
+constexpr int kPreTransposeMinorDim = 2;
+constexpr int kPreTransposeVectorizedDim = 3;
+
+int GetVectorSize(const HloFusionAnalysis& analysis,
+                  absl::InlinedVector<int64_t, 4> input_dims,
+                  Vector3 permutation, int64_t threads_per_block) {
+  int vector_size = 2;
+  int64_t threads_max = analysis.device_info().core_count() *
+                        analysis.device_info().threads_per_core_limit();
+
+  auto threads_counts = [&](int vec_size) {
+    int64_t blocks = 1;
+    for (int i = 0; i < permutation.size(); ++i) {
+      if (i == permutation.size() - 1) {
+        blocks *=
+            CeilOfRatio(input_dims[permutation[i]], WarpSize() * vec_size);
+        continue;
+      }
+      if (permutation[i] == input_dims.size() - 1) {
+        blocks *=
+            CeilOfRatio(input_dims[permutation[i]], WarpSize() * vec_size);
+        continue;
+      }
+      blocks *= input_dims[permutation[i]];
+    }
+    return blocks * threads_per_block;
+  };
+
+  while (vector_size != 1) {
```

Comment:
This is a bit confusing. Looks like this loop will run at most once?

```
// Enable vectorization iff the below conditions are met:
// (1) The minor dim of pre transpose and post transpose should be
// divisible by warp_size * vector_size. (2) Device occupancy is high.
if (is_aligned && thread_counts(2) >= threads_max)) {
  return 2;
}
return 1; 
```

Also, since vec_size is now a constant in the lambda, we can just make it a real constant.